File: live/python/general_/optional.py
from mpack.option import Option, ty
import random

class Employee:
    def __init__(self, name: str) -> None:
        self.name = name

    def greet(self) -> None:
        print(f"Hello {self.name}, how was your day?")
    
    def __str__(self) -> str:
        return f'Employee(name={self.name !r})'


def get_employee() -> ty.Optional[Employee]:
    _names: list[str] = ["Simon", "Nganga", "Njoroge"]
    name = random.choice(_names)
    return None if random.random() < 0.5 else Employee(name)


def chname(name: str):
    def _chname(pe: Employee | None) -> Employee | None:
        if pe is None: return
        return Employee(name)

    return _chname

class ContextLike(ty.Protocol):
    def _enter(self) -> ty.Any: ...
    def _exit(self) -> ty.Any: ...

def Context(obj: ContextLike):
    try:
        val = obj._enter()
        yield val or obj
    finally:
        obj._exit()

class _OpenFile:
    def __init__(self, file: 'File') -> None:
        self.file = file
        self.content = ''
    
    def write(self, text):
        self.content += str(text)
    
    def read(self):
        return self.content

class File:
    def __init__(self, file: str) -> None:
        self.filename = file
    
    def _enter(self) -> _OpenFile:
        print("Opening file:", self.filename)
        return _OpenFile(self)
    
    def _exit(self):
        print("Closing file:", self.filename)

for of in Context(File("hello.txt")):
    print("In scope of:", of.file.filename)
    of.write("Hello World, how are your?")
    # raise FileNotFoundError
    print("File contents:", of.read())

print()

me = Option(get_employee()).value_or(Employee("Faith Njeri"))
print(me)
me.greet()

print()

emp = Option(get_employee())
emp.and_then(lambda e: e.greet())\
    .or_else(lambda: print("Invalid Employee"))\
    .transform(chname("Lydia Njeri"))\
    .and_then(lambda e: e.greet())\
    .or_else(lambda: print("Still Invalid"))

File: live/javascript/optional.js
import { Option } from './option.mjs'

class Employee {
  constructor(name) {
    this.name = name;
  }
  greet() {
    console.log(`Hello ${this.name}, how was your day?`);
  }
  toString() {
    return `Employee(name='${this.name}')`;
  }
}

function* Context(obj) {
  if (!obj) throw TypeError("Expected a non-null value, got", obj);
  try {
    let val = undefined
    if (obj.__enter__) val = obj.__enter__();
    yield val || obj;
  } finally {
    if (obj.__exit__()) obj.__exit__();
  }
}

class File {
  constructor(filename) {
    this.filename = filename
    this.content = ''
  }
  write(content) {
    this.content += content
  }
  read() {
    return this.content
  }
  __enter__() {
    console.log("Opening file:", this.filename)
  }
  __exit__() {
    console.log("Closing file:", this.filename)
  }
}

function get_employee() {
  let _names = ["Simon", "Nganga", "Njoroge"];
  let name = _names[Math.round(Math.random() * 100) % _names.length];
  return Math.random() < 0.5 ? new Employee(name) : undefined;
}

let me = new Option(get_employee());

me.and_then((e) => e.greet())
  .or_else((_) => console.log("Invalid Employee"))
  .transform((e) => (e ? new Employee(e.name.toLowerCase()) : null))
  .and_then((e) => e.greet())
  .or_else((_) => console.log("invalid employee"));

for (let f of Context(new File("hello.txt"))) {
  console.log("In scope of:", f.filename)
  f.write("Hello World")
  // throw "Custom Error"
  console.log("File contents:", f.read())
}